### PR TITLE
Redesigned Plugin Catalog 

### LINKED ISSUE #1067: Sectioning plugin catalog
With growing number of plugins of different types our Plugin Catalog may benefit from sectioning "Maintained plugins" in PLUGIN_CATALOG.md. How about creating a section for each plugin type, for example like below, as opposed to having all plugins on a single list?

## Maintained collector plugins
| Name  | Description | Link | Download |
| :---- | :---------- | :--- | :------- |
| CollectorPlugin| Plugin description | [snap-plugin-collector-apache](https://github.com/intelsdi-x/snap-plugin-collector-apache) | [Linux](https://s3-us-west-1.amazonaws.com/snap-plugins-linux-latest/snap-plugin-collector-apache)

## Maintained processor plugins
| Name  | Description | Link | Download |
| :---- | :---------- | :--- | :------- |
| ProcessorPlugin | Plugin description| [snap-plugin-collector-apache](https://github.com/intelsdi-x/snap-plugin-collector-apache) | [Linux](https://s3-us-west-1.amazonaws.com/snap-plugins-linux-latest/snap-plugin-collector-apache)

## Maintained publisher plugins
| Name  | Description | Link | Download |
| :---- | :---------- | :--- | :------- |
| PublisherPlugin | Plugin description| [snap-plugin-collector-apache](https://github.com/intelsdi-x/snap-plugin-collector-apache) | [Linux](https://s3-us-west-1.amazonaws.com/snap-plugins-linux-latest/snap-plugin-collector-apache)

### LINKED ISSUE #1138: Add process to add plugins to plugin catalog
We need to outline how people add to the plugin catalog in our CONTRIBUTING.md file. To date it's been that someone adds a line to the PLUGIN_CATALOG.md and we scrape from that, but I know @nanliu is working on automating it w/a JSON object instead. Let's get something out there sooner than later. cc @sarahjhh

### PR DESCRIPTION
**Pull request by mjbrender**
Fixes #1067
Fixes #1138 

Redesigned Plugin Catalog to list all plugins by collector, processor or publisher. Link describes whether it's under development or not and `Maintainer` generates based on URL. 

This file will be generated automatically based on our packaging tools in the next rev or 2. I added some information in manually that can be generated next time. 

@intelsdi-x/snap-maintainers

### CODE REVIEW

**Comment by kindermoumoute** [comment]
I quite like @andrzej-k idea to reduce the link text to its plugin name instead of having the full path.
https://github.com/intelsdi-x/snap/issues/1067#issue-165515196

**Comment by mjbrender** [comment]
Good catch @kindermoumoute - thank you! I corrected it.

**Comment by candysmurf** [comment]
👍 , LGTM

**Comment by mjbrender** [comment]
Thanks for reviewing all - @nanliu and I are working on programmatically adding the download URLs as well, so holding before publishing.

File: docs/plugins.yml [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+- intelsdi-x/snap-plugin-publisher-heka
+- intelsdi-x/snap-plugin-publisher-influxdb
+- intelsdi-x/snap-plugin-publisher-kafka
+- intelsdi-x/snap-plugin-publisher-kairosdb
+- intelsdi-x/snap-plugin-publisher-mysql
+- intelsdi-x/snap-plugin-publisher-opentsdb
+- intelsdi-x/snap-plugin-publisher-postgresql
+- intelsdi-x/snap-plugin-publisher-rabbitmq
+- intelsdi-x/snap-plugin-publisher-riemann
+# @mjbrender: I'm not sure if brharrington wants to release this officially
```
Comment (IRCody) [inline docs/plugins.yml]: Should this comment stay in?

File: docs/PLUGIN_CATALOG.md [hunk 2]
```diff
@@ -92,5 +94,4 @@ There will always be more plugins we wish we had. To make sure others can contri
 | Collectd (native) - [Issue needed](https://github.com/intelsdi-x/snap/issues/)        | Collector | Compatibility for legacy environments |                                                                                          |
 | Prometheus    - [#1196](https://github.com/intelsdi-x/snap/issues/1196)               | Publisher | As part of a CNCF stack               |                                                                                          |
 | Kubernetes Node (Minion) - [Issue needed](https://github.com/intelsdi-x/snap/issues/) | Collector | Monitoring nodes in k8s               |                                                                                          |
-| JVM (via JMX) - [Issue needed](https://github.com/intelsdi-x/snap/issues/)            | Collector |                                       |                                                                                          |
 | Ceilometer - [Issue needed](https://github.com/intelsdi-x/snap/issues/)               | Publisher | possibly just OSLO proc + RMQ         |                                                                                          |
```
Comment (IRCody) [inline docs/PLUGIN_CATALOG.md]: Is the wish list meant to be in any sort of order?

File: docs/PLUGIN_CATALOG.md [hunk 3]
```diff
@@ -1,85 +1,87 @@
...
+| Procnum | Collector | raintank | No description. | [snap-plugin-collector-procnum](https://github.com/raintank/snap-plugin-collector-procnum) | [GitHub](https://github.com/raintank/snap-plugin-collector-procnum/releases/latest) |
+| Procstat | Collector | Staples-Inc | Collect process metrics using the snap telemetry framework | [snap-plugin-collector-procstat](https://github.com/Staples-Inc/snap-plugin-collector-procstat) | [GitHub](https://github.com/Staples-Inc/snap-plugin-collector-procstat/releases/latest) |
+| RabbitMQ | Collector | Intel | Collects RabbitMQ metrics from management API | [snap-plugin-collector-rabbitmq](https://github.com/intelsdi-x/snap-plugin-collector-rabbitmq) | [GitHub](https://github.com/intelsdi-x/snap-plugin-collector-rabbitmq/releases/latest) |
+| Redfish | Collector | Intel | Collects metrics from Redfish API  | In development by [@candysmurf](https://github.com/candysmurf) | No release date - [please open an issue](https://github.com/intelsdi-x/snap/issues) if you'd like to see this |
```
Comment (IRCody) [inline docs/PLUGIN_CATALOG.md]: is the In development by added manually or via the script somehow?

File: docs/PLUGIN_CATALOG.md [hunk 4]
```diff
@@ -1,85 +1,87 @@
...
+| Smart | Collector | Intel | Collects Intel SSDs S.M.A.R.T. metrics | [snap-plugin-collector-smart](https://github.com/intelsdi-x/snap-plugin-collector-smart) | [GitHub](https://github.com/intelsdi-x/snap-plugin-collector-smart/releases/latest) |
+| Swap | Collector | Intel | Collects Linux swap metrics from /proc | [snap-plugin-collector-swap](https://github.com/intelsdi-x/snap-plugin-collector-swap) | [GitHub](https://github.com/intelsdi-x/snap-plugin-collector-swap/releases/latest) |
+| Users | Collector | Intel | Collects Linux user metrics from utmp | [snap-plugin-collector-users](https://github.com/intelsdi-x/snap-plugin-collector-users) | [GitHub](https://github.com/intelsdi-x/snap-plugin-collector-users/releases/latest) |
 
-## Committed plugins
-These plugins are in planned/active development. This list is useful if you want to reach out and contribute to the development.
+### Processors
+| Name  | Type  | Maintainer | Description | Link | Download |
+| :---- | :---- | :--------- | :---------- | :--- | :------- |
```
Comment (IRCody) [inline docs/PLUGIN_CATALOG.md]: Collectors have a release column while processors/publishers have a download column. Is there a reason for the difference?

File: docs/plugins.yml [hunk 5]
```diff
@@ -0,0 +1,78 @@
+---
+#####################
+# We're excited you wrote plugin! Here's how you share it with others:
+#
+# 1. Add a newline anywhere in the list below
+# 2. List your plugin following the <user>/<repo> format
```
Comment (IRCody) [inline docs/plugins.yml]: Do we do some sort of processing on the repo name (i.e. looking for snap-plugin-{collector,processor,publisher}-x in order to put it into the right section? What happens if I add a plugin like `ircody/super-cool-plugin`?

If it requires the repo to be named something specifically maybe it's worth calling out.

File: docs/PLUGIN_CATALOG.md (hunk 2 above)
Comment (mjbrender) [inline docs/PLUGIN_CATALOG.md]: Not to date, but it'd be nice if it did. Any recommendations?

File: docs/PLUGIN_CATALOG.md (hunk 3 above)
Comment (mjbrender) [inline docs/PLUGIN_CATALOG.md]: Manually - what's your thought on it? The reason I kept it is that it was listed previously.

File: docs/plugins.yml (hunk 5 above)
Comment (mjbrender) [inline docs/plugins.yml]: @IRCody good call. Our plugin authoring documentation has the naming convention down, but it's worth mentioning here.

File: docs/PLUGIN_CATALOG.md (hunk 2 above)
Comment (IRCody) [inline docs/PLUGIN_CATALOG.md]: I was thinking by plugin type {collector->process->publisher} and within those groups could be alphabetical.

File: docs/PLUGIN_CATALOG.md (hunk 3 above)
Comment (IRCody) [inline docs/PLUGIN_CATALOG.md]: I was just thinking that maintaining that manually would be potentially error prone. If we are going to list stuff in development maybe it makes sense to have a way to call that out inside the file we are using to generate the list and have a standard way of representing it.

**Comment by mjbrender** [comment]
Thanks for all the comments @IRCody. To answer your questions: 

> Would it make sense to have the Maintainer column be a link to the user/org that is attached to that repo?
Great improvement - I can add this in now. The go-forward direction is to have configuration information in the repository that includes a specific maintainer we can link to. For now we can assume the org/user is the right one. 

> Could the Name and link columns be joined so that the name is also a link to the repo? Is there a reason this is bad?
Let's do that. 

The one additional - we'd like to offer our test binaries for convenience for the ones we have. It will be the `latest_build`, while releases will be the latest tag. 

Updates coming.

File: docs/plugins.yml (hunk 5 above)
Comment (nanliu) [inline docs/plugins.yml]: Currently the order matters. If we just sort alphabetically, ircody/... would be higher on the list than intelsdi-x/... So would it be better to follow the order as it appears here so you have some control?

Comment (mjbrender) [inline docs/plugins.yml]: I would like to continue to sort by the name of the plugin, not the maintainer. In this particular markdown file, we will not include the maintainer in the name as part of the sort.

**Comment by mjbrender** [comment]
Based on talking with @nanliu - this PR is a v1 we will follow up with v2 once the files land and we can continue to developer the helper utility. I'll clean up the language points you brought up @IRCody and wait on redesign of the table for v2.

File: docs/plugins.yml (hunk 5 above)
Comment (IRCody) [inline docs/plugins.yml]: >Currently the order matters. If we just sort alphabetically, ircody/... would be higher on the list than intelsdi-x/... So would it be better to follow the order as it appears here so you have some control?

How do we differentiate between process/publisher then? It looks at the name of the plugin right? It also seems like it'd be pretty easy to sort on what comes after the `/`?

Comment (nanliu) [inline docs/plugins.yml]: Yeah, we can do sort by name per Matt's suggestions.

**Comment by mjbrender** [comment]
If you're good with it @IRCody, I'll go with this for now.

**Pull request merged**